procfs/proc.go

```diff
@@ -34,40 +34,75 @@ type Procer interface {
 	NewProcProc() ([]ProcProc, error)
 }
 
-//NewProcProc collects data from various proc pseudo-file system files
-//and converts it into a ProcProc structure.
+var (
+	processCPUTally      = map[int]uint64{}
+	previousTotalCPUTime = totalCPUTime()
+)
+
+// NewProcProc collects data from various proc pseudo-file system files
+// and converts it into a ProcProc structure.
 func NewProcProc() ([]ProcProc, error) {
-	procs, err := procfs.AllProcs()
+	allProcs, err := procfs.AllProcs()
 	if err != nil {
 		return []ProcProc{}, err
 	}
 
-	var ps []ProcProc
+	var (
+		output             = []ProcProc{}
+		newProcessCPUTally = map[int]uint64{}
+		newTotalCPUTime    = totalCPUTime()
+	)
 
-	for _, proc := range procs {
+	for _, proc := range allProcs {
 		cli, err := proc.CmdLine()
 		if err != nil || len(cli) == 0 {
 			continue
 		}
 
-		var p ProcProc
-		p.CmdLine = cli
-
-		p.PID = proc.PID
-
-		comm, _ := proc.Comm()
-		p.Comm = comm
+		comm, err := proc.Comm()
+		if err != nil {
+			continue
+		}
 
 		stat, err := proc.NewStat()
 		if err != nil {
-			continue // because the rest of the values can't be queried
+			continue
 		}
 
-		p.VirtualMemory = stat.VirtualMemory()
-		p.ResidentMemory = stat.ResidentMemory()
-		p.CPUTime = stat.CPUTime()
+		var utilization float64
+		newProcCPUTime := uint64(stat.UTime + stat.STime)
+		newProcessCPUTally[proc.PID] = newProcCPUTime
+		if _, exists := processCPUTally[proc.PID]; exists {
+			utilization = float64(newProcCPUTime-processCPUTally[proc.PID]) / float64(newTotalCPUTime-previousTotalCPUTime)
+		}
 
-		ps = append(ps, p)
+		output = append(output, ProcProc{
+			CmdLine:        cli,
+			PID:            proc.PID,
+			Comm:           comm,
+			VirtualMemory:  stat.VirtualMemory(),
+			ResidentMemory: stat.ResidentMemory(),
+			CPUUtilization: utilization,
```

Comment:
Yes, and that's the case on the first cycle, before we've stored anything about that PID in the map. Instead of reporting goofy numbers, we go with `0` instead.